File: app/src/main/java/com/salino/sali - Copy/data/remote/source/RecurringRemoteDataSource.kt
package com.salino.sali.data.remote.source

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.Query
import com.salino.sali.data.model.RecurringItem
import kotlinx.coroutines.tasks.await
import javax.inject.Inject

class RecurringRemoteDataSource @Inject constructor(
    private val firestore: FirebaseFirestore
) {
    private fun recurringCollection(householdId: String) = firestore.collection("households")
        .document(householdId)
        .collection("recurringItems")

    fun listenToRecurringItems(householdId: String, onItems: (List<RecurringItem>) -> Unit): ListenerRegistration {
        return recurringCollection(householdId)
            .orderBy("updatedAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, _ ->
                val items = snapshot?.documents?.mapNotNull { doc ->
                    doc.toObject(RecurringItem::class.java)?.copy(id = doc.id)
                } ?: emptyList()
                onItems(items)
            }
    }

    suspend fun upsertRecurringItem(householdId: String, item: RecurringItem) {
        recurringCollection(householdId).document(item.id).set(item).await()
    }

    suspend fun deleteRecurringItem(householdId: String, recurringItemId: String) {
        recurringCollection(householdId).document(recurringItemId).delete().await()
    }
}

File: ios/Remote/RecurringRemoteDataSource.swift
import Foundation
import FirebaseFirestore

final class RecurringRemoteDataSource {
    private let firestore: Firestore?

    init(firestore: Firestore? = configuredFirestore()) {
        self.firestore = firestore
    }

    private func recurringCollection(_ householdId: String) throws -> CollectionReference {
        guard let firestore else { throw FirebaseConfigurationError.missingGoogleServiceInfo }
        return firestore.collection(FirestoreCollections.households)
            .document(householdId)
            .collection(FirestoreCollections.recurringItems)
    }

    func listenToRecurringItems(householdId: String, onItems: @escaping ([RecurringItem]) -> Void) -> ListenerRegistration {
        guard let collection = try? recurringCollection(householdId) else {
            onItems([])
            return NoopListenerRegistration()
        }
        return collection
            .order(by: "updatedAt", descending: true)
            .addSnapshotListener { snapshot, _ in
                let items = snapshot?.documents.map { RecurringItem(documentId: $0.documentID, data: $0.data()) } ?? []
                onItems(items)
            }
    }

    func upsertRecurringItem(householdId: String, item: RecurringItem) async throws {
        try await recurringCollection(householdId).document(item.id).setDataAsync(item.firestoreData)
    }

    func deleteRecurringItem(householdId: String, recurringItemId: String) async throws {
        try await recurringCollection(householdId).document(recurringItemId).deleteAsync()
    }
}
